### PR TITLE
stdinc: Document some APIs; remove SDL_RESTRICT in favor of doc comments

### PR DESCRIPTION
**Pull request by castholm**
## Description

I found some spare time to document a few of the SDL_stdinc.h APIs ~~and add missing SDL_RESTRICT qualifiers to match the libc function signatures.~~

~~I also added `SDL_RESTRICT` qualifiers to the corresponding implementations. Let me know if you believe there are any problems with doing this.~~

Update: The `SDL_RESTRICT` macro has now been removed entirely in favor of documenting aliasing invariants in doc comments.

### CODE REVIEW

File: include/SDL3/SDL_stdinc.h
Comment (castholm) [inline include/SDL3/SDL_stdinc.h:426]: I removed this since there's no `SDL_swscanf`/`SDL_vswscanf`.

File: include/SDL3/SDL_stdinc.h
Comment (castholm) [inline include/SDL3/SDL_stdinc.h:1402]: I want to bring attention to the behavior of SDL's fallback implementations of `SDL_wcstol` and similar APIs, which I don't think match the behavior of most libc implementations. I didn't explicitly test them, but from just reading the code many of them appear to be subtly wrong. For example, `wcstol` supposed to support bases between 1 and 36, and specifying 0 should detect `0`-prefixed octal and `0x`/`0X`-prefixed, but SDL's implementation appears to only correctly support bases 1 to 10 and 16, and 0 only accepts lowercase `0x`.

Comment (castholm) [reply include/SDL3/SDL_stdinc.h:1402]: These mismatches between the libc functions and SDL's fallbacks make it difficult to correctly document these APIs, since it's not clear whether the difference in behavior is a bug or an intentional subset of the libc specs.

Comment (madebr) [reply include/SDL3/SDL_stdinc.h:1402]: Good catch!

This is a bug.
SDL's stdinc function aim to be a complete replacement of libc.
If you feel up to the job, a (failing) unit test in `test/testautomation_stdlib.c` would be appreciated.
And if you're still with us, you might be able to fix this.
(If you don't, we can jump in)

Comment (castholm) [reply include/SDL3/SDL_stdinc.h:1402]: I could try adding a few failing tests (if I do, should I disable them with `TEST_DISABLED` so that builds still pass?), but I don't think I would have the time/patience to fix all the behavior and and edge cases on my own. I could open a follow-up tracking issue instead. Especially with things like `SDL_strtod` and all the ways float parsing can go wrong it seems like it would require a quite significant amount of effort to achieve correctness.

Comment (slouken) [reply include/SDL3/SDL_stdinc.h:1402]: In general SDL's functions aim to be a functional subset of the C library functionality, not a complete implementation. In general we add what is commonly needed by games, and if it's missing an edge case or functionality we add it as needed by real-world use cases.

Comment (madebr) [reply include/SDL3/SDL_stdinc.h:1402]: Creating a new issue suffices :)

**Comment by slouken** [comment]
@icculus, how much do we want SDL_RESTRICT in the API definitions? Does it actually provide meaningful speed increase? I'm weighing readability vs performance here. I think having it in memcpy(), etc. probably makes sense, but we don't need this everywhere in the API.

@castholm, I don't think you need SDL_RESTRICT in the function implementations, just the signatures. You can double check this, but that's the way other attributes have worked.

**Comment by castholm** [comment]
`restrict` is a type qualifier, so my understanding is that if you actually want the compiler to be able to take advantage of it for optimizations, you need to also specify it in the implementation, not just the prototype. Compare the outputs in this [Godbolt demo](https://godbolt.org/z/qvq3br8ff).

In function prototypes `restrict` appears to do nothing on its own, but speaking from the point of view of a user I would still prefer having them in the headers, since they limit the number of assumptions the programmer can make. E.g. if I see an unannotated `SDL_utf8strlcpy`, I might incorrectly assume that it's safe to pass overlapping buffers to it even though its implementation might call the libc `memcpy` and corrupt the results.

(The main reason I've been going through this header in first place is because I've been hand-crafting some SDL3 language bindings and want the downstream bindings to be correctly annotated with regard to aliasing, non-nullness, null-termination, etc., so I've had to dig around a bit in both libc docs and the SDL implementations to answer certain questions.)

File: include/SDL3/SDL_stdinc.h
Comment (castholm) [reply include/SDL3/SDL_stdinc.h:1402]: #10750

**Review by slouken — APPROVED** [review]

**Comment by maia-s** [comment]
It'd be helpful if functions using restrict stated something to that effect in the documentation. The restrict qualifier often gets lost when translating to other programming languages. I'm considering adding some code to my own generator to add this to generated documentation, but I don't know how common that is in general.

File: src/stdlib/SDL_strtokr.c
Comment (slouken) [inline src/stdlib/SDL_strtokr.c:24]: I don't think this is correct. A common pattern is to use strtok() on a pointer and fill in the return value with the advanced pointer.

File: src/stdlib/SDL_string.c
Comment (slouken) [inline src/stdlib/SDL_string.c:1217]: Again, same thing here, often the same pointer is passed into both arguments.

**Comment by slouken** [comment]
Please remove SDL_RESTRICT from functions unless it's performance critical or the underlying implementation requires it.

**Comment by icculus** [comment]
Yeah, I don't love adding SDL_RESTRICT to things but the documentation is really good (and I'm happy someone else did this work instead of me!)

File: src/stdlib/SDL_strtokr.c
Comment (castholm) [reply src/stdlib/SDL_strtokr.c:24]: It is correct. The memory pointed to by `ptr` should not overlap with either of `s1` or `s2`. Did you look at the pages I linked in the PR description? [Here's musl's implementation of the function if you want another reference](https://github.com/bminor/musl/blob/dd1e63c3638d5f9afb857fccf6ce1415ca5f1b8b/src/string/strtok_r.c).

**Comment by castholm** [comment]
> Please remove SDL_RESTRICT from functions unless it's performance critical or the underlying implementation requires it.

@slouken Please clarify, what do you mean exactly? If you're compiling with `HAVE_STRTOK_R` etc. config macros set so that `SDL_strtok_r` is just `return strtok_r(s1, s2, ptr)`, you probably want the qualifiers for correctness, no? I didn't just haphazardly add the qualifiers where I thought they would fit, I did it to match what libc does.

If the idea is that you don't care for `restrict` at all because it just muddies up the signatures, then  I would suggest we remove `SDL_RESTRICT` altogether in favor of just documenting the "must not overlap" invariants in the doc comments. Only using it for some libc wrappers but not all doesn't make a lot of sense, and only using it in the headers but not the implementations makes it amount to nothing more than glorified documentation for parameters.

File: include/SDL3/SDL_stdinc.h
Comment (castholm) [inline include/SDL3/SDL_stdinc.h:642]:
```suggestion
 * If `size` is 0, it will be set to 1.
```
Note to self that #10756 changed the behavior and made it well-defined.

File: src/stdlib/SDL_strtokr.c
Comment (castholm) [reply src/stdlib/SDL_strtokr.c:24]: (To clarify, the `restrict` applies to the first level of indirection (the variable to which the pointer to the substring will be written), not the string itself.)

**Comment by slouken** [comment]
> If the idea is that you don't care for `restrict` at all because it just muddies up the signatures, then I would suggest we remove `SDL_RESTRICT` altogether in favor of just documenting the "must not overlap" invariants in the doc comments.

Let's go ahead and do this. We can re-add SDL_RESTRICT later if we think it's important, without breaking the ABI.

File: include/SDL3/SDL_stdinc.h
Comment (smcv) [inline include/SDL3/SDL_stdinc.h:635]: This is a bit misleading: the returned pointer might *be* `mem`. Perhaps this?

```suggestion
 * If this function returns a non-NULL pointer not equal to `mem`,
 * the `mem` pointer will have been freed and cannot be dereferenced anymore.
```

File: include/SDL3/SDL_stdinc.h
Comment (smcv) [reply include/SDL3/SDL_stdinc.h:642]: #10756 has been merged, so I think it would be better to apply this change.

Perhaps it would be worth adding something like this (from #10769):

```
 * In some `realloc` implementations, `realloc(mem, 0)` frees `mem`
 * without allocating any memory to replace it. `SDL_realloc` does not
 * do this: `SDL_realloc(mem, 0)` always attempts to allocate and return
 * a memory region, so its behaviour is consistent with `SDL_realloc(mem, 1)`.
```

**Comment by castholm** [comment]
Updates after the last force push:

- Removed `SDL_RESTRICT`.
- Clarified behavior of `SDL_realloc` as per @smcv's suggestions.
- Documented a few additional APIs (strcpy, strtol, etc.); please re-audit them for correctness.

File: include/SDL3/SDL_stdinc.h
Comment (smcv) [inline include/SDL3/SDL_stdinc.h:1264]:
```suggestion
 * If you are confident that the regions never overlap, using SDL_memcpy() may improve performance.
```

I'm sure randomly zeroing memory will improve performance sometimes, so technically what you said is true, but I don't think it's what you meant :-)

File: include/SDL3/SDL_stdinc.h
Comment (smcv) [inline include/SDL3/SDL_stdinc.h:1668]:
```suggestion
 * The result of calling `SDL_atof(str)` is equivalent to `SDL_strtod(str, NULL)`.
```

**Comment by smcv** [comment]
(drive-by review only, not thorough)

File: include/SDL3/SDL_stdinc.h
Comment (slouken) [inline include/SDL3/SDL_stdinc.h:1538]: This was a bug, fixed in 80e541d1f.

File: include/SDL3/SDL_stdinc.h
Comment (slouken) [inline include/SDL3/SDL_stdinc.h]: I don't think that's true here. It's totally valid to call this:
```c
SDL_strtol(str, &str, 10);
```

File: include/SDL3/SDL_stdinc.h
Comment (slouken) [reply include/SDL3/SDL_stdinc.h:1539]: The return value change was intentional, but I see that we're not actually checking the return value anywhere, so we can change this to match SDL_strlcpy(), for consistency.

File: include/SDL3/SDL_stdinc.h
Comment (castholm) [reply include/SDL3/SDL_stdinc.h]: I reworded these after your review to clarify that the aliasing restriction applies the address to which the pointer to the substring will be written, not the string itself, which is correct and how the libc functions are defined. There's no aliasing happening in `SDL_strtol(str, &str, 10)`.

If you think that information is redundant and want it removed it's your call; I felt like keeping it so that people writing bindings for languages with better type systems than C can extract all the information they need from the documentation alone.

File: include/SDL3/SDL_stdinc.h
Comment (slouken) [reply include/SDL3/SDL_stdinc.h:1539]: Actually, we return the number of bytes written because the actual number can be shorter than both SDL_strlen(src) and dst_bytes, if src ends with a truncated UTF-8 sequence.

Comment (castholm) [reply include/SDL3/SDL_stdinc.h:1539]: Yeah I also just realized that as well. So returning `dst_bytes` makes sense, but it's still worth mentioning this difference so that people don't replace calls to `strlcpy` with `utf8strlcpy` and end up with subtle bugs.

Comment (slouken) [reply include/SDL3/SDL_stdinc.h:1539]: Yep, agreed.

File: include/SDL3/SDL_stdinc.h
Comment (slouken) [reply include/SDL3/SDL_stdinc.h:1680]: I'd like to leave it in, as it's pretty convenient and what is often used in example code around the net.

File: include/SDL3/SDL_stdinc.h
Comment (slouken) [reply include/SDL3/SDL_stdinc.h:1751]: The name of these should probably be changed and the implementation guarantee that 64-bit values are supported.

Comment (slouken) [reply include/SDL3/SDL_stdinc.h:1751]: Actually, long long is at least 64-bits on all supported platforms. I think we just need to compile-time assert that and we can leave this alone.

Comment (slouken) [reply include/SDL3/SDL_stdinc.h:1751]: I'll change the return types to long long though, for consistency with the C API.

Comment (slouken) [reply include/SDL3/SDL_stdinc.h:1751]: (after your PR lands, so you don't have to resolve conflicts)

**Review by slouken — APPROVED** [review]

File: include/SDL3/SDL_stdinc.h
Comment (slouken) [inline include/SDL3/SDL_stdinc.h:1489]: What is `dst`?

File: include/SDL3/SDL_stdinc.h
Comment (castholm) [reply include/SDL3/SDL_stdinc.h:1750]: :+1:
There's also `SDL_lltoa`/`SDL_ulltoa` but those currently call `_i64toa`/`_ui64toa` so they are not quite the same. But maybe they should be renamed for clarity.

File: include/SDL3/SDL_stdinc.h
Comment (castholm) [reply include/SDL3/SDL_stdinc.h]: Updated.

File: include/SDL3/SDL_stdinc.h
Comment (castholm) [reply include/SDL3/SDL_stdinc.h]: Whoops, copy/paste typo, fixed now.

File: include/SDL3/SDL_stdinc.h
Comment (slouken) [inline include/SDL3/SDL_stdinc.h:1489]: I'm still not sure what this means. Can you give me an example where this constraint is violated?

Comment (castholm) [reply include/SDL3/SDL_stdinc.h]: I ultimately ended up removing any mention of aliasing from these comments.

My understanding was that it's to guard against programs like these where the string data and the location the output address will be written to both overlap:
```c
#include <stdlib.h>
#include <stdio.h>

int main(void) {
    char buf[50] = "0123456789abcdef";

    char *s = buf;
    char **t = (char **)(&buf[8]);

    // any value written to '*t' will affect what 's' sees
    long long result = strtoll(s, t, 10);

    printf("result: %lld\n", result);
    printf("s: %p %s\n", s, s);
    printf("t: %p\n", *t);

    return 0;
}

// result: 1234567
// s: 0000001050BFFC70 01234567xⁿ┐P►
// t: 0000001050BFFC78
```
"Writing to `*t` must not what we can read from `s`!" is what the qualifier is trying to say, but this is already such a convoluted scenario it's unlikely to show up in actual code, plus it might already be UB, I'm not sure. Having it in the docs is probably only going to cause confusion. I googled around a bit and found a few discussions from mailing lists where even Gcc people seem to be confused about the meaning of `restrict` in `strtol` and similar functions.

Comment (slouken) [reply include/SDL3/SDL_stdinc.h]: Ah, okay, that makes sense. I think removing the mention was the right call.

**Comment by slouken** [comment]
Merged, thanks!

**Pull request merged**
